### ISSUE TITLE
Use the training process to inform the data loading process

### ISSUE THREAD

**Issue by xiamengzhou** [mosaicml/streaming#347]
Hi team, thanks for your fantastic work on creating all these open-sourced repos! I have been using them for my projects and it has been an amazing experience. I am implementing an idea that tries to dynamically load data from different sources, where I maintain shard information for each dataset and a sampling ratio to decide which dataset to get sample_ids from. The following is the part of the `__iter__` function on how to get the next sample:

```
while True:
  sampling_ratio = self.sampling_ratio
  domain = np.random.choice(range(self.n_domains), 1, p=sampling_ratio)[0].item()
  domain_sample_id = sample_ids[domain][self.domain_id[domain] % domain_lens[domain]]
  self.domain_id[domain] += 1
  yield self[(domain, domain_sample_id)]
```

I've rewritten `__getitem__`, `load_state_dict()`, `state_dict()` etc. functions to make it compatible. One issue is that I need to change the sampling_ratio periodically based on the evaluation loss, and I have been doing it in a callback function each time after the evaluation is done by directly changing the sampling_ratio through state.train_dataloader.dataset. However, this procedure does not work because each worker of `__iter__` does not share memory. I am thinking about using SharedMemory to hold this information and allow the main process to write it and the workers to read it, using a combination of event and lock. Do you think this would work? Is it possible to write the shared memory using a nonworker process and use the worker processes to access it?

For simplicity, I can also only use the main process to load data and it seems to be working now.

Thanks for your help!

**Comment by karan6181** [comment]
Hi @xiamengzhou, what happens if you hold a object variable such as `sampling_ratio` in your dataset class, which I think you are doing based on this line `sampling_ratio = self.sampling_ratio` !! Now, to change the `sampling_ratio`, I think you can directly call `self.sampling_ratio = new_sampling_ratio` from anywhere in your code and your `num_workers` would read the new value. Let me know if I misunderstood your question completely.

**Comment by xiamengzhou** [comment]
With this implementation, the `self.sampling_ratio` is not updated in the workers unless I use a single thread for loading data. In the multi-worker case, each worker seems to have their own copy of the dataset object, and setting the ratio does not change the value of the workers. So I thought that I would need to use shared memory to put in this parameter to be accessed by different workers?

**Comment by karan6181** [comment]
ohh I see. Yes in that case, you can use the shared memory to access the data (create a list of length num_workers maybe) from the workers. You just have to be careful on which worker fetches which data !!

**Comment by karan6181** [comment]
@xiamengzhou Checking if were able to make it work or not!

**Comment by xiamengzhou** [comment]
Hi, thanks for following up with it! I decided to use a single worker in the end to reduce the coding complexity. Fortunately, the throughput was not affected at all. I will reopen the issue if I get into any issues when implementing the multi-process solution :) Thanks a ton for your help!
